### PR TITLE
fix(config): Allow for fields starting with '@'

### LINKED ISSUE #4273: Lookup error caused by field prefixed with "@"
### Vector Version

```
nightly from oct 2
```

### Vector Configuration File

```toml
[transforms.add_fields]
	type = "add_fields"
	inputs = ["input"]
	overwrite = false
	fields."@timestamp" = "{{ timestamp }}"
```

### Expected Behavior

Happily parse a field starting with "@"

### Actual Behavior
```
Oct 02 20:55:18.418  INFO vector::app: Log level "info" is enabled.
Oct 02 20:55:18.419  INFO vector::app: Loading configs. path=["/etc/vector/vector.toml"]
Oct 02 20:55:18.448 ERROR vector::topology: Configuration error: Transform "add_fields":  --> 1:1
  |
1 | @timestamp
  | ^---
  |
  = expected lookup
```

I discussed with @Hoverbear briefly and should have a PR in shortly.

### PR DESCRIPTION
**Pull request by spencergilbert**
Closes #4273

"@timestamp" is frequently (always?) used in the ELK stack and supporting this field would be nice for users transitioning from ELK.

Signed-off-by: Spencer Gilbert <spencer.gilbert@gmail.com>

### CODE REVIEW

**Review by bruceg — COMMENTED** [review]
This looks straightforward enough, but I'm a little leery of adding this. This change would making using `@` in any other place in the syntax messy at best. Can these field names be quoted effectively?

**Comment by spencergilbert** [comment]
> This looks straightforward enough, but I'm a little leery of adding this. This change would making using `@` in any other place in the syntax messy at best. Can these field names be quoted effectively?

do you mean in the config?

```toml
fields."@timestamp" = 
```

If so that's what's currently breaking in my previously working config file :(

**Comment by spencergilbert** [comment]
Confirmed that double quotes, single quotes, no quotes, and back ticks don't work (nor some random attempts at escaping)

**Comment by spencergilbert** [comment]
Any more thoughts on this @bruceg @JeanMertz ?

**Comment by Hoverbear** [comment]
Hmmm, I wonder if we should only allow this inside quoted strings then?

**Review by Hoverbear — COMMENTED** [review]
This is the right fix for this specific issue, but I think we need to be careful about how this impacts the greater system.

**Comment by binarylogic** [comment]
Shouldn't we solve quoting? That seems like the real problem here.

**Comment by Hoverbear** [comment]
@binarylogic do you think that

```toml
fields."@timestamp" = "foo"
```

Is not the valid solution here? What would you propose?

**Comment by binarylogic** [comment]
Yep, I do. I'd be very surprised if that doesn't work, we have quoting examples in the RFC:

https://github.com/timberio/vector/blob/master/rfcs/2020-07-21-2744-remap-syntax.md#guide-level-proposal

**Comment by Hoverbear** [comment]
Right, the existing remap/lookup parser does not handle that. Our friend @spencergilbert is exploring how to fix this so that we'll conform to the RFC!

**Comment by spencergilbert** [comment]
@Hoverbear @binarylogic @bruceg do we want/need to ensure @timestamp would only be valid when contained by quotes?

**Comment by binarylogic** [comment]
I'll defer to @JeanMertz and @FungusHumungus, but anything not in the reserved characters list should be fine. I'm sure they'll have opinions on if we want this to be strict to avoid any future clashes if they syntax expands.

**Comment by spencergilbert** [comment]
I confirmed that with the change Vector will error when the `@timestamp` field is unquoted.
```
root@darter:/git/timberio/vector# vim /tmp/test.toml
root@darter:/git/timberio/vector# ./target/release/vector validate /tmp/test.toml
√ Loaded ["/tmp/test.toml"]
√ Component configuration
√ Health check `out`
---------------------------
                  Validated
root@darter:/git/timberio/vector# vim /tmp/test.toml
root@darter:/git/timberio/vector# ./target/release/vector validate /tmp/test.toml
Failed to load ["/tmp/test.toml"]
---------------------------------
x unexpected character found: `@` at line 8

root@darter:/git/timberio/vector# cat /tmp/test.toml
[sources.in]
	type = "stdin"

[transforms.add_fields]
	type = "add_fields"
	inputs = ["in"]
	overwrite = false
	fields.@timestamp = "{{ timestamp }}"

[sinks.out]
	inputs = ["add_fields"]
	type = "console"
	encoding.codec = "text"
```

**Comment by FungusHumungus** [comment]
The problem of quoting fields within remap is it will then become ambiguous as to whether you are referring to a string or a field.
It will not be clear, for example, if you call:

```
split(.foo, "@field")
```

whether you want to split foo by the value of a field called `@field` or by the text `"@field"`.

We could make a special case such that if your string begins with a `@`, then it is actually a field. But I don't like that idea.

I can't think of a technical reason why we would need to quote a field containing an @ at all in remap, although there is then an inconsistency between the remap syntax and the fields in the rest of the config.

Would we be able to quote fields with a different character to the " we use for strings?

**Comment by JeanMertz** [comment]
> The problem of quoting fields within remap is it will then become ambiguous as to whether you are referring to a string or a field.

This isn't accurate.

To start, there are two terminologies to keep in mind:

* `path` — contains a series of rules to go from the root of the event to a nested field
* `segment` — one piece of the path to go from the current event level to a child-level

Paths _always_ start with a dot (`.`), optionally followed by one or more segments, and each segment is separated by another `.`. So whether path segments are quoted or not, has no impact on the parser's ability to tell a path from a string, since a string starts with a quote, and a path with a dot.

In your example, you'd do the following:

```rust
split(.foo, "@field") // split by string value
split(.foo, ."@field") // split by field value
```

You should think of quoted path segments as the canonical way to reference fields within an event.

Quoted path segments should allow _any valid character_ (where "valid" is defined as whatever we consider acceptable, but should basically mean any valid UTF-8 character). The reason we use quotes, is so that we can feed the parser a pair of tokens (`"`) by which it can determine the start and end of a path segment, and consider all characters in-between as part of the segment.

The "naked-path" option we support (e.g. `.foo.bar`) exists solely for convenience. It is there so that people don't constantly have to write `."foo"."bar"`, which would become tedious. However, since we don't have a pair of delimiter tokens to determine when the path starts or ends, we have to set some constraints, which in our case means the syntax has to adhere to this rule:

```peg
path_field_name = { (ASCII_ALPHA | "_" | "-" )+ }
```

At this point, all characters defined in this rule are off-limits for other use-cases within the context of a path. For example, `[` isn't allowed here because we use that to reference path indexes (e.g. `.foo[1]` gets the second element in an array at path `.foo`), which is why @bruceg brought up his concern.

---

Back to the original topic, @spencergilbert, @Hoverbear and @bruceg are all three correct:

1. We want to support `@timestamp`, for sure!
2. We _might_ want to avoid supporting `.@timestamp` since that would prevent us from using `@` for any other syntax in the future.
3. We _must_ support `."@timestamp"` to make this use-case work, and thus we should make sure the `inner_quoted_string` rule handles this as expected.

I think the best way forward for now is to make sure **3** works, since that's clearly an oversight if that's currently not the case, given the example @spencergilbert gave (`@timestamp`) is a fairly common one.

I'm also somewhat inclined to think that `.@timestamp` should be supported as well, since — as mentioned — the `@` sign is often used in services that add their own fields and want to "namespace" those fields in some way to prevent overriding existing fields. But I think we can hold off on that for now to first consider the long-term implications, as long as quoted path segments support this use-case.

**Comment by binarylogic** [comment]
@JeanMertz I agree. Can you update or close this PR? If you close it, can you make sure we have the appropriate issues open and assigned for the next sprint?

**Comment by JeanMertz** [comment]
@binarylogic, @spencergilbert and I chatted on Discord, they are going to add a test case for this, and then we'll know if it's an issue with the PEG definition or the parser itself (I expect it's the latter).

We can then move forward to a solution to make the third case work as expected.

@spencergilbert as I mentioned on Discord; I'm happy to take this over from you if you want, but I don't want to be in your way, so I'll just leave you to it, unless you tell me otherwise.

**Comment by spencergilbert** [comment]
@JeanMertz I added a unit test to the add_fields behavior test which is failing on master (passes on 0.10.0) which reflects the regression I noticed.

**Comment by Hoverbear** [comment]
@JeanMertz Nice digging. 😏

**Comment by JeanMertz** [comment]
@spencergilbert I pushed some changes to this PR that should fix the described issue.

The changes in this PR won't allow bare `@timestamp`, but it does correctly parse `"@timestamp"` now.

There are still some tasks left to do, but those can be separated from this PR, namely:

* improve error messages for invalid paths
* share more logic between `Lookup` and the remap parser
* a couple of helper functions to reduce parsing complexity
* (optionally) allow bare `@` character in path lookup, meaning `.foo.@timestamp` would work, instead of `.foo."@timestamp"`

**Comment by spencergilbert** [comment]
Awesome! Sounds perfect

**Comment by binarylogic** [comment]
@JeanMertz can you open issue for those things are do you plan to solve them here?

File: src/mapping/parser/grammar.pest
Comment (Hoverbear) [inline src/mapping/parser/grammar.pest:5]: 😻

File: src/transforms/add_fields.rs
Comment (Hoverbear) [inline src/transforms/add_fields.rs:121]: Gosh here you are doing the next step of https://github.com/timberio/vector/blob/master/rfcs/2020-05-25-2692-more-usable-logevents.md :)

File: src/transforms/add_fields.rs
Comment (Hoverbear) [inline src/transforms/add_fields.rs:128]: This forces another `.to_string()` on top of the `into()` call above and is not ideal. This is why line 105 existed. :)

File: src/transforms/add_fields.rs
Comment (Hoverbear) [inline src/transforms/add_fields.rs:133]: Similarly, this forces a reallocation and has performance consequences, and is the reason for L105 existing. :)

**Review by Hoverbear — CHANGES_REQUESTED** [review]
The `pest` and `lookup` changes look good but the new usage of `Lookup` in some transforms is not ideal due to the early stage of this new type.

**Comment by binarylogic** [comment]
@JeanMertz can you add a behavior test for the `add_fields`, `rename_fields`, and `remove_fields` transforms? That should ensure special characters work as expected in those.

**Comment by FungusHumungus** [comment]
I'm not sure I fully understand why the changes to the pest file were necessary. Code like the following currently works:

```
."@ook" = upcase(.onk)
.erk."@ook" = downcase(."@ook")
```

Do you have an example of some code that doesn't work now that this fixes?

**Comment by JeanMertz** [comment]
@FungusHumungus I wrote down what currently works and doesn't work in:

https://github.com/timberio/vector/pull/4277#issuecomment-707771117

This PR does three things:

1. It adds a test showing how to actually write quoted paths in TOML (e.g. the need for nested quotes)
2. It fixes the issue of those nested quotes ending up in the final event (e.g. no more `{ "\"timestamp\"": "..." }`
3. While fixing this, I noticed I had to fix it in multiple places due to converging pathing logic since we introduced the `Lookup` rule. The Pest file changes consolidates pathing into one set of rules, used throughout Vector.

**Comment by Hoverbear** [comment]
@jamtur01 I'd like this to block 0.11 if it could.

**Comment by jamtur01** [comment]
@Hoverbear Okay. I don't have an issue here.

@JeanMertz Can you please update? Thanks.

**Comment by JeanMertz** [comment]
I'm working on getting this in with the latest remap-lang changes, should be good to go by tomorrow.

**Comment by jamtur01** [comment]
@JeanMertz What's the status here?

**Comment by JeanMertz** [comment]
@jamtur01 `Lookup` work got reverted due to some issues we discovered, so this PR is no longer valid right now, but will have to be applied again once we merge the lookup work back into master again.

I'll close this for now, but will make sure to tackle this when I focus on the lookup work in the next couple of weeks.

**Comment by Hoverbear** [comment]
Thank you so much @spencergilbert for creating this. :)

**Pull request closed without merging**
